llvm/lib/Target/RISCV/RISCVVMV0Elimination.cpp

```diff
@@ -0,0 +1,154 @@
+//===- RISCVVMV0Elimination.cpp - VMV0 Elimination -----------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===---------------------------------------------------------------------===//
+//
+// Mask operands in vector pseudos have to be in v0. We select them as a virtual
+// register in the singleton vmv0 register class instead of copying them to $v0
+// straight away, to make optimizing masks easier.
+//
+// However the register allocator struggles with singleton register classes and
```

Comment:
This is a hand wavy and poor explanation for the problem and makes it sound like a workaround. The problem is primarily with every other piece of code that can now produce situations where multiple values must be live at the same time

Reply:
To my knowledge, it is a workaround?  I don't understand your second sentence.  All the words individually make sense, I just don't get the connection to the point.  I've never really understood why 3 virtual registers with overlapping live ranges in a 2 element register class is fine, but 2 overlapping virtual registers in a 1 element register class isn't.  

Also, do you have better wording to use here?  Or is this more than a request to reword a comment?

Reply:
> I've never really understood why 3 virtual registers with overlapping live ranges in a 2 element register class is fine, but 2 overlapping virtual registers in a 1 element register class isn't.

Maybe this is just a limitation of how GreedyRegAlloc implements spilling, i.e. it requires at least two registers in a class?

One way to think of RISCVVMV0Elimination is that it "spills" vmv0 register classes to vr register classes by inserting copies and inflating the register class. That way we're removing any overlapping live ranges.

Reply:
The problem is if you have more than 1 value that need to be live at the same point, as in instruction.

```
  %0:my_rc = DEF0
  %1:my_rc = DEF1
  SOME_INST implicit %0, implicit %1
```
Is the kind of case that will fail. Other cases where they don't need to be live at the same point should work

Reply:
Off the top of my head the reason why we were ending up with instructions with two separate uses of vmv0 was because copies were getting coalesced into the vmv0 class. There shouldn't be any instructions with two vmv0 operands.

It might be worth trying to change this pass to inflate any unnecessary vmv0 register classes so we avoid this situation? Ideally in a separate PR so we minimise the test diff from isel's current behaviour

Reply:
Luke, I'm not clear what you mean by "this pass".  It sounds like you're referring to the VMV0 elimination pass from this patch, but that doesn't seem to make sense in context?  Were you maybe referring to Register Coalescing?

Reply:
My idea is that RISCVVMV0Elimination sits just before register allocation, after register coalescer. But instead of emitting copies to V0, it finds instructions with two or more distinct uses of VMV0.

Only one of the uses should need to be in VMV0, I.e the mask operand.

For the other uses, it would either inflate the uses to VR, or if not possible emit a COPY to VR. 

That way each instruction would have at most one use of VMV0, and this avoids the scenario that @arsenm was mentioning, which should be allocatable. 

I'm suggesting we could explore it as a follow up to this patch.